[IV-24-25] Objetivo 4

abbonno opened the pull request:
* [x]  ¿Hay un camino claro que permita ir desde el código hasta la historia de
   usuario que desarrolla, pasando por el mensaje de commit y el issue
   correspondiente?
* [x]  ¿Se ha intentado cubrir con tests una parte de la lógica de negocio que se
   esté desarrollando?
* [x]  ¿Hay un producto mínimamente viable que describa lo que se va a entregar en
   este objetivo?
* [x]  ¿En este producto mínimamente viable se han priorizado unas clases,
   módulos o paquetes más fundamentales frente a otros, movidos a otro
   PMV/Hito?
* [x] ¿Estoy leyendo todo lo que marco o simplemente marco lo que se me pone por
delante?
Sí, estoy leyendo
* [x] ¿Se han documentado las elecciones de biblioteca de aserciones y del
   de *test runner*?
* [x] ¿Te has asegurado que lo que mencionas como biblioteca de aserciones y
test runners lo son realmente y tienen la misma funcionalidad? ¿O estás poniendo
`unittest` como *test runner* y `pytest` como biblioteca de aserciones?
* [x] ¿Esa documentación incluye criterios de aceptación y también criterios de
búsqueda para las diferentes opciones?
* [x]  ¿Se han seguido los principios F.I.R.S.T. en la creación del test y
   se ha documentado cómo se ha hecho?
Fast: son tests rápidos (unos 2 milisegundos)
Independent: ninguno depende de otra función testeada
Repeatable: los datos usados son invariables, da el mismo resultado cada vez
Self-validating: aserciones claras que muestran cuando falla, en caso contrario el test se pasa correctamente
Timely: los tests son oportunos ya que estamos siguiendo TDD y es a partir de ellos que generamos la lógica de negocio correcta.
* [x]  ¿Se describe claramente en el PMV los tests que hay que pasar para que se
   considere viable?
* [x] ¿Se han testeado también las excepciones?

abbonno committed:
Escoge y documenta su uso (biblioteca de aserciones) #20

Se decide emplear el estándar de golang

abbonno committed:
Escoge y documenta su uso (test runner) #21

Se emplea el estándar de golang

abbonno committed:
Crea lógica de negocio y tests, completa documentación #22

abbonno committed:
Merge branch 'main' into Objetivo4

juanbarearojo commented on internal/turnos-asignados_test.go:
```diff
@@ -0,0 +1,142 @@
+)
+
+func TestNuevoTurno(t *testing.T) {
+	// Datos de entrada
```
Las buenas prácticas nos hablan de que el código no debería tener comentarios ya que esto a largo plazo genera deuda técnica. ¿Has considerado hacer un código más claro para que no sea necesario?

abbonno replied:
[79c7eef](https://github.com/abbonno/healthScheduler/pull/23/commits/79c7eef79274a220c37acda7d2eab0f617eb0ead)

juanbarearojo commented on internal/turnos-asignados_test.go:
```diff
@@ -0,0 +1,142 @@
+	turno := NuevoTurno(nombre, area, fecha)
+
+	// Aserciones
+	if turno.Nombre != nombre {
```
Lo test deberían ser unitarios. ¿Por que no le das un vuelta para que cada test evalue una única cosa?

abbonno replied:
Esta es la manera más unitaria que he encontrado de organizar mis tests.
[00fd866](https://github.com/abbonno/healthScheduler/pull/23/commits/00fd86617ba742f451478bdeab9d6eca727299b1)
[72b9145](https://github.com/abbonno/healthScheduler/pull/23/commits/72b914592539c01873e719b10e26092182cf7ddb)
[4053bed](https://github.com/abbonno/healthScheduler/pull/23/commits/4053bedb8cf73235b9ba8b3394ff8573ed8adcee)

juanbarearojo commented on internal/turnos-asignados_test.go:
```diff
@@ -0,0 +1,142 @@
+	fecha := time.Now()
+
+	// Llamada al constructor
+	turno := NuevoTurno(nombre, area, fecha)
```
Si tu constructor no tiene ninguna condicón dentro de él ¿Crees que es necesario testear que se crea el objeto? Es raro que se produzca un error al crearlo

abbonno replied:
[79c7eef](https://github.com/abbonno/healthScheduler/pull/23/commits/79c7eef79274a220c37acda7d2eab0f617eb0ead)

juanbarearojo commented on internal/turnos-asignados.go:
```diff
@@ -15,17 +17,68 @@ const (
+	fechaActual := time.Now()
+
+	// Generar turnos para los próximos 365 días
+	for i := 0; i < 365; i++ {
```
¿Crees que es recomendable el uso de números mágicos?

abbonno replied:
Tienes razón, gracias por la corrección!

abbonno replied:
[79c7eef](https://github.com/abbonno/healthScheduler/pull/23/commits/79c7eef79274a220c37acda7d2eab0f617eb0ead)

juanbarearojo commented on internal/turnos-asignados.go:
```diff
@@ -15,17 +17,68 @@ const (
+
+func GenerarPlanSemanal(enfermeros []Empleado) (TurnosAsignados, error) {
+	totalEmpleados := len(enfermeros)
+	if totalEmpleados < 10 {
```
¿Has pensado que el uso de número mágicos puede disminuir la legibilidad de tu código?

abbonno replied:
[79c7eef](https://github.com/abbonno/healthScheduler/pull/23/commits/79c7eef79274a220c37acda7d2eab0f617eb0ead)

hossam1522 commented on Taskfile.yml:
```diff
@@ -4,4 +4,8 @@ tasks:
     desc: Validar la sintaxis de los archivos fuente de Go.
     cmds:
-      - gofmt -e internal/turnos-asignados.go 
+      - go vet ./...
```
¿Este cambio qué tiene que ver con este objetivo?

abbonno replied:
Al hacer el pull request ha habido un conflicto y se ha incluido la línea anterior cuando ya estaba cambiada a la nueva. No comprendo por qué ha ocurrido, disculpa.

hossam1522 replied:
No hombre no te disculpes xd

JJ replied:
... todavía tienes que corregirlo. Has resuelto mal el conflicto.

juanbarearojo commented on internal/turnos-asignados_test.go:
```diff
@@ -0,0 +1,142 @@
+	// Primera verificación: que no haya áreas duplicadas
+	for _, turnos := range plan {
+		for _, turno := range turnos {
+			if turnosPorDia[turno.Fecha] == nil {
```
¿Crees que es recomendable que es un test evalue tantas cosas? En general sirven para depurar si fallara este test en que parte habría dado fallo

abbonno replied:
Tienes razón, debería replantear la estructura de mis tests. Lo haré a continuación.

abbonno replied:
Esta es la manera más unitaria que he encontrado de organizar mis tests.
[00fd866](https://github.com/abbonno/healthScheduler/pull/23/commits/00fd86617ba742f451478bdeab9d6eca727299b1)
[72b9145](https://github.com/abbonno/healthScheduler/pull/23/commits/72b914592539c01873e719b10e26092182cf7ddb)
[4053bed](https://github.com/abbonno/healthScheduler/pull/23/commits/4053bedb8cf73235b9ba8b3394ff8573ed8adcee)

juanbarearojo commented on docs/biblioteca_aserciones.md:
```diff
@@ -0,0 +1,30 @@
+
+Se emplearán los siguientes criterios en la búsqueda y selección de una de las bibliotecas de aserciones empleadas en nuestro proyecto.
+
+- Mantenimiento: se actualiza frecuentemente.
```
¿Has considerado mirar la puntuaciones de estas herramientas en snyk advisor?

abbonno replied:
[01072fe]

abbonno replied:
He hecho lo mismo para las herramientas de test runner. [fb37b99]

abbonno commented:
@JJ

hossam1522 commented on internal/turnos-asignados.go:
```diff
@@ -15,17 +17,68 @@ const (
+	indiceEnfermero := 0
+	fechaActual := time.Now()
+
+	// Generar turnos para los próximos 365 días
```
Poner comentarios implica que tu código no es lo suficientemente claro como para entenderse sin ellos. Eso implica deuda técnica.

abbonno replied:
Es cierto, gracias por el apunte!

abbonno replied:
[79c7eef](https://github.com/abbonno/healthScheduler/pull/23/commits/79c7eef79274a220c37acda7d2eab0f617eb0ead)

juanbarearojo commented:
Buenas he revisado tu PR y he dejado un par de comentarios para que puedas mejorarlo. Mucho ánimo

hossam1522 commented on internal/turnos-asignados.go:
```diff
@@ -15,17 +17,68 @@ const (
+	return turnosAsignados, nil
+}
+
+func GetTurnosEnfermero(enfermero *Empleado, turnosAsignados TurnosAsignados) []Turno {
+	return turnosAsignados[enfermero]
+}
```
¿Realmente hace falta esta función?

abbonno replied:
[79c7eef](https://github.com/abbonno/healthScheduler/pull/23/commits/79c7eef79274a220c37acda7d2eab0f617eb0ead)

abbonno replied:
He optado por no usarla finalmente. Para mostrar por pantalla los turnos de cada enfermero se puede acceder directamente con el argumento dado. [f237f1e](https://github.com/abbonno/healthScheduler/pull/23/commits/f237f1edb9e306de7f13a42fda510c3db01e13a0)

hossam1522 commented on internal/turnos-asignados_test.go:
```diff
@@ -0,0 +1,142 @@
+)
+
+func TestNuevoTurno(t *testing.T) {
+	// Datos de entrada
```
Lo dicho con los comentarios.

abbonno replied:
[79c7eef](https://github.com/abbonno/healthScheduler/pull/23/commits/79c7eef79274a220c37acda7d2eab0f617eb0ead)

hossam1522 commented on internal/turnos-asignados.go:
```diff
@@ -15,17 +17,68 @@ const (
 type TurnosAsignados map[*Empleado][]Turno
 
-func NuevoTurno(nombre TipoTurno, area string, fecha time.Weekday) *Turno {
+func NuevoTurno(nombre TipoTurno, area string, fecha time.Time) *Turno {
 	return &Turno{
 		Nombre: nombre,
 		Area:   area,
 		Fecha:  fecha,
 	}
 }
```
¿No viene un constructor por defecto para las estructuras de Go?

abbonno replied:
Tienes razón, con new.

abbonno replied:
[79c7eef](https://github.com/abbonno/healthScheduler/pull/23/commits/79c7eef79274a220c37acda7d2eab0f617eb0ead)

hossam1522 commented on internal/turnos-asignados_test.go:
```diff
@@ -0,0 +1,142 @@
+	// Segunda verificación: que no haya áreas sin cubrir
+	for dia, turnos := range turnosPorDia {
+		for turno, areas := range turnos {
+			if len(areas) < 10 {
```
Me he perdido, ¿de dónde viene este 10?

abbonno replied:
Se trata del total de áreas, por lo que es el mínimo de empleados que debe existir para poder cubrirlas todas. Arreglaré el error de uso de números mágicos.

abbonno replied:
[79c7eef](https://github.com/abbonno/healthScheduler/pull/23/commits/79c7eef79274a220c37acda7d2eab0f617eb0ead)

abbonno replied:
[79c7eef](https://github.com/abbonno/healthScheduler/pull/23/commits/79c7eef79274a220c37acda7d2eab0f617eb0ead)

hossam1522 commented on docs/biblioteca_aserciones.md:
```diff
@@ -0,0 +1,30 @@
+
+Se trata de una librería moderna de aserciones entre las herramientas de gotest.tools, pero tiene un mantenimiento un poco inestable.
+
+### Estándar (con el paquete testing)
+
+La propia herramienta de testing estándar de Golang ofrece la posibilidad de realizar validaciones.
```
Validaciones que no aserciones. No puedes definir esto como biblioteca de aserciones.

abbonno replied:
[01072fe]

hossam1522 reviewed (commented):
Aparte de lo comentado, decirte que tengas cuidado ya que tienes muy pocos commits para la cantidad de código que has introducido.

juanbarearojo commented:
También te recomiendo que antes de corregir más código o mas revisiones le des una vuelta a los issues y a la metodología que has seguido. Ten en cuenta que la metodología de este objetivo es muy parecida a la del 2

abbonno committed:
Añade el parámetro health de snyk Advisor como indicador objetivo de mantenimiento #21

abbonno committed:
Añade el parámetro health de snyk Advisor como indicador objetivo de mantenimiento y corrige redacción #20

abbonno committed:
Soluciona malas prácticas en el código #22

Se eliminan números mágicos, comentarios innecesarios, constructor y otras funciones innecesarias

abbonno committed:
Merge branch 'Objetivo4' of github.com:abbonno/healthScheduler into Objetivo4

abbonno committed:
Recupera estado anterior código #22

Deshacemos cambios en el código para reestructurarlos y que se muestren adecuadamente en los issues

abbonno committed:
Crea TestAreaOcupadaDoblemente #26

Este test comprueba que no haya más de un enfermero asignado al mismo área al mismo tiempo en el plan generado

abbonno committed:
Crea TestAreaSinOcupar #25

Comprueba que no falte nadie en un área en cualquier momento

abbonno committed:
Crea TestMostrarTurnosEnfermero #27

Comprueba que la información sobre sus turnos se muestra correctamente

abbonno committed:
Crea GenerarPlanAnual #24

Crea el planning de turnos para todo un año dados los enfermeros contratados

abbonno committed:
Crea MostrarTurnosEnfermero #27

Convierte los turnos asignados a un enfermero en string y lo devuelve

abbonno committed:
Soluciona error argumentos MostrarTurnosEnfermero #27

abbonno commented:
> Aparte de lo comentado, decirte que tengas cuidado ya que tienes muy pocos commits para la cantidad de código que has introducido.

He intentado volver a dar pasos sobre el código para documentar los cambios en los cambios realizados. Adicionalmente, he creado sus issues correspondientes, espero que sean suficientes. Si tienes alguna recomendación es bienvenida. Muchas gracias!

abbonno commented:
> Buenas he revisado tu PR y he dejado un par de comentarios para que puedas mejorarlo. Mucho ánimo

Gracias por tu colaboración Juan. He reorganizado un poco la estructura de los tests y cambiado los issues. Espero tú revisión,

hossam1522 commented on internal/turnos-asignados.go:
```diff
@@ -15,17 +17,59 @@ const (
-		Area:   area,
-		Fecha:  fecha,
+const totalAreas = 10
+const diasAnio = 365
```
Hm. ¿Y si es bisiesto?

abbonno replied:
También lo he pensado, pero lo que se está generando realmente es un plan del momento en el que se ejecuta a un año por delante o sea, 365 días. En realidad, este es un parámetro arbitrario, ya que en la historia de usuario no se especifica con respecto a cuánto tiempo por delante se quiere saber los turnos ¿Una semana? ¿Un mes? He decidido un año porque da margen al enfermero.

hossam1522 replied:
Entonces simplemente ponle otro nombre. Lo del año confunde.

hossam1522 commented on internal/turnos-asignados_test.go:
```diff
@@ -0,0 +1,109 @@
+	areas := []string{"Área1", "Área2", "Área3", "Área4", "Área5", "Área6", "Área7", "Área8", "Área9", "Área10"}
+	tiposTurno := []TipoTurno{Mañana, Tarde, Noche}
+
+	// Crear un mapa para verificar las áreas ocupadas
```
Aquí seguimos teniendo comentarios en el código.

abbonno replied:
Lo he modificado en un commit pero no sé por qué no aparece. Aún así está solucionado en el siguiente también, puedes comprobarlo: [1fa43cf](https://github.com/abbonno/healthScheduler/pull/23/commits/1fa43cfe08434be0151985a9bb762b7b3ccca617)

wickeet commented on internal/turnos-asignados.go:
```diff
@@ -15,17 +17,59 @@ const (
+	fechaActual := time.Now()
+
+	for i := 0; i < diasAnio; i++ {
+		fecha := fechaActual.AddDate(0, 0, i)
```
¿Esto no crea todas las fechas con la hora en el momento que se crea el Plan Anual? No sé yo como de correcto es que alguien del turno de noche tenga que su fecha es el 05 de febrero a las 15:37. A no ser que el atributo hora del tipo de dato sea irrelevante, en ese caso igual sería mejor usar otro tipo de dato, ¿no crees? Algo como un struct con un dia, un mes y un año (puedes usar time.Month y time.Year) y crear tu propio AddDate y no tendrías que cambiar apenas el código que ya tienes hecho.

JJ replied:
Y aquí sí que voy a parar. Tenéis que hacer modificaciones como respuesta a *todos los comentarios*. Cerrar un comentario sin ningún tipo de comentario ni modificación es una falta de respeto a la persona que lo ha hecho y ha invertido su tiempo en ello.

wickeet reviewed (commented):
Solucionando lo que han dicho los compañeros y esta puntualización yo lo veo bastante bien

abbonno committed:
Elimina comentarios

evaanngiil commented on docs/biblioteca_aserciones.md:
```diff
@@ -0,0 +1,26 @@
+Se emplearán los siguientes criterios en la búsqueda y selección de una de las bibliotecas de aserciones empleadas en nuestro proyecto.
+
+- Mantenimiento: se actualiza frecuentemente. También se empleará la plataforma ["snyk Advisor"](https://snyk.io/advisor/golang) que mide el grado de "health" o salud de una herramienta dado su estado de mantenimiento.
+- Estándar: en caso de que el anterior factor no sea claramente decisivo, se optará por elegir la herramienta que cumple el estándar del lenguaje elegido.
```
¿Qué significa ¨cumple el estándar del lenguaje elegido¨?

abbonno replied:
Se refiere a las herramientas por defecto que ofrece el lenguaje, lo que al igual que al decidir el uso de otras herramientas se considera una buena práctica. En el caso de go existe el comando "go test".

JJ replied:
Si @evaanngiil te ha preguntado algo, una respuesta *nunca* resuelve el comentario y no deberías haberlo hecho. Él mismo te ha dicho que :+1: , pero todavía lo tengo que revisar yo y me encuentro un comentario *que no está resuelto con un commit*.
En todo caso, "el estándar del lenguaje" no significa nada. Di con precisión qué es lo que quieres decir.

evaanngiil commented on docs/biblioteca_aserciones.md:
```diff
@@ -0,0 +1,26 @@
+
+### [Gomega](https://github.com/onsi/gomega)
+
+Gomega se apoya en Ginkgo para su uso, por lo que ambos proyectos se mantienen a la par (se actualiza frecuentemente). Se coloca en la última revisión con una alta puntuación "health" de 96 sobre 100.
```
Solamente esta opción cuenta con una valoración en la plataforma elegida. Puedes tratar de elegir otro factor para medir el mantenimiento de la herramienta (Github)

abbonno replied:
Justamente ese es el primer medidor que se está usando. Como puedes observar se ofrece el enlace al repositorio de github en el título de la herramienta analizada. Esto se hace porque es la plataforma en la que se obtiene la información "Gomega se apoya en Ginkgo para su uso...". Sobre la plataforma snyk advisor, aunque solo encontremos información de esta herramienta, pienso que es adecuado tenerlo en cuenta y no prescindir de ello.

evaanngiil replied:
No prescindir, pero si hacer referencia por ejemplo al número de commits en el ultimo año o factores que indiquen que la herramienta es mantenida.

evaanngiil commented on internal/turnos-asignados_test.go:
```diff
@@ -0,0 +1,109 @@
+}
+
+func TestAreaSinOcupar(t *testing.T) {
+	enfermeros := []Empleado{
+		"Enfermero1", "Enfermero2", "Enfermero3", "Enfermero4", "Enfermero5",
+		"Enfermero6", "Enfermero7", "Enfermero8", "Enfermero9", "Enfermero10",
+		"Enfermero11", "Enfermero12", "Enfermero13", "Enfermero14", "Enfermero15",
+	}
```
Trata de evitar repetición de código para seguir las buenas prácticas, si algo se repite muchas veces seria conveniente declararlo 1 sola vez y reutilizarlo.

abbonno replied:
Tienes razón, ahora mismo lo soluciono.

evaanngiil commented on internal/turnos-asignados_test.go:
```diff
@@ -0,0 +1,109 @@
+	// Crear un mapa para verificar las áreas ocupadas
+	ocupacion := make(map[time.Time]map[TipoTurno]map[string]bool)
+
+	// Recorrer los turnos asignados y marcar las áreas como ocupadas
+	for _, turnos := range turnosAsignados {
+		for _, turno := range turnos {
+			if ocupacion[turno.Fecha] == nil {
+				ocupacion[turno.Fecha] = make(map[TipoTurno]map[string]bool)
```
Deberías usar el principio SOLID para desarrollar los tests, este nos dice que cada test debería ser independiente de los otros y realizar una sola cosa.

abbonno replied:
Comprendo SOLID pero no comprendo bien cómo lo incumple en este caso. Lo único que se está comprobando es que no haya ningún área vacía. De hecho soluciona el anterior test, que unía las dos comprobaciones sobre el plan anual, lo que sí se alejaba del "Separation of concerns". Si pudieras aclarar tu punto de vista te lo agradecería.

evaanngiil replied:
En el guión se informa, si un test tiene mas de un if, está haciendo varias tareas y por lo tanto no es correcto. Te lo detallo más. Las funciones de prueba TestAreaOcupadaDoblemente y TestAreaSinOcupar realizan múltiples tareas. Esto dificulta aislar la causa de un fallo. Si una prueba falla, ¿es un problema con la generación del plan o con la lógica de verificación dentro de la prueba? Además, hace que las pruebas sean más complejas de entender y mantener.

evaanngiil reviewed (changes requested):
Buenas, se me ha asignado revisar tu PR. Te he puesto algunos comentarios con ánimo de que mejores y entiendas los conceptos claves del objetivo. Ánimo.

juanmaaf commented on internal/turnos-asignados.go:
```diff
@@ -15,17 +17,59 @@ const (
+	for i := 0; i < diasAnio; i++ {
+		fecha := fechaActual.AddDate(0, 0, i)
+
+		for _, tipoTurno := range tiposTurno {
```
¿Consideras que la mejor solución para resolver el problema que se quiere resolver es anidar for e if? Te invito a revisar mi PR y verás que me mencionan la complejidad ciclomática. No conozco GO y no sé si implementa filters, pero considera usar maps, como me recomiendan.

juanmaaf commented on internal/turnos-asignados_test.go:
```diff
@@ -0,0 +1,109 @@
+
+	turnosPorDia := make(map[time.Time]map[TipoTurno]map[string]bool)
+
+	for _, turnos := range plan {
```
Igual que en el comentario que he realizado arriba.

juanmaaf reviewed (commented):
Buenas. Además de los comentarios realizados por otros compañeros, que veo que no se han resuelto, te dejo unas puntualizaciones para reflexionar y considerar en el código.

abbonno committed:
Resolución de fallos tests #25

abbonno committed:
Cambia nombre confuso #24

Un año puede durar 365 o 366 días por lo que es incorrecto llamar a la constante de esa manera

abbonno committed:
Crea clase Fecha y los test pertinentes #28

abbonno committed:
Adapta código a la nueva clase fecha #28

abbonno committed:
Soluciona DRY código

abbonno committed:
Añade referencia al número de commits como parámetro de mantenimiento #20

abbonno committed:
Añade referencia al número de commits como parámetro de mantenimiento #21

abbonno committed:
Crea el tipo de dato Ocupación

Este dato es empleado en la comprobación de las áreas ocupadas cada día y turno

hossam1522 reviewed (approved):
No se me ocurre más que comentar.

juanmaaf reviewed (approved):
Espero que todos los comentarios te hayan ayudado para superar el objetivo. Por mi parte tienes el apto.

abbonno commented:
@JJ

JJ commented on docs/biblioteca_aserciones.md:
```diff
@@ -0,0 +1,26 @@
+
+### [Paquete gotest.tools/assert](https://github.com/gotestyourself/gotest.tools)
+
+Se trata de una librería moderna de aserciones entre las herramientas de gotest.tools, pero tiene un mantenimiento un poco inestable. Prueba de ello es que solo se han realizado 23 commits este mes en un intervalo de 3 días seguidos. No existe una entrada para puntuar la herramienta en snyk Advisor, por lo que no se puede juzgar mediante este parámetro.
```
Por favor, reescribe esa frase de forma que se entienda lo que quieres decir. En todo caso, "mantenimiento inestable" no tiene nada que ver con loscriterios que has puesto arriba.

JJ commented on docs/biblioteca_aserciones.md:
```diff
@@ -0,0 +1,26 @@
+
+## SELECCIÓN FINAL
+
+Siguiendo los criterios se optará por emplear el estándar de Golang. La propia herramienta de testing estándar de Golang ofrece la posibilidad de realizar validaciones, suficientes para realizar la función de las aserciones.
```
¿De qué herramienta de testing estamos hablando? ¿Cuál de las tres que has incluido arriba?

JJ commented on docs/test_runner.md:
```diff
@@ -0,0 +1,30 @@
+
+### [Go testing package](https://pkg.go.dev/testing)
+
+Se trata del paquete estándar de golang por lo que es preferente.
```
Paquete estándar ¿para qué? ¿En qué sentido es estándar?

JJ commented on docs/test_runner.md:
```diff
@@ -0,0 +1,30 @@
+Se emplearán los siguientes criterios en la búsqueda y selección de un test runner adecuado para nuestro proyecto.
+
+- Mantenimiento: se trata de una herramienta que se actualiza a menudo respecto a errores o nuevas necesidades. También se empleará la plataforma ["snyk Advisor"](https://snyk.io/advisor/golang) que mide el grado de "health" o salud de una herramienta dado su estado de mantenimiento.
+- Estándar: en caso de que el anterior factor no sea claramente decisivo, se optará por elegir la herramienta que cumple el estándar del lenguaje elegido.
```
De veras, no sigas hablando de estándares. Un estándar es una norma reconocida de alguna forma, con algún documento. Hay algo que se llama "librería estándar" que es el conjunto de bibliotecas, que no son parte del lenguaje, pero que *se distribuyen habitualmente con el lenguaje*. Pero ni cumplen ni dejan de cumplir ningún estándar que diga "Se considera Go el lenguaje que incluya esta cosa o la otra". Simplemente es parte del API del lenguaje.

JJ commented on docs/test_runner.md:
```diff
@@ -0,0 +1,30 @@
+
+## SELECCIÓN FINAL
+
+Siguiendo los criterios elegidos anteriormente, escogeremos usar el paquete estándar de go.
```
¿Que se llama?

JJ reviewed (changes requested):
NO has dicho nada sobre la herramienta CLI que vas a usar. Por favor, repasa el guión.

abbonno committed:
Soluciona la documentación de la biblioteca de aserciones elegida #20

abbonno committed:
Soluciona la documentación del test runner elegido #21

abbonno committed:
Documenta la selección del CLI a emplear #29

abbonno committed:
Corrección comando task check #19

Se corrige el comando elegido para comprobar la sintáxis de los archivos fuente, debido a un error en la corrección

abbonno commented:
@JJ

JJ commented on docs/biblioteca_aserciones.md:
```diff
@@ -5,7 +5,7 @@
 
 - Mantenimiento: se actualiza frecuentemente. Esto se puede observar en los datos observados en el apartado "Insights" de los repositorios en github. También se empleará la plataforma ["snyk Advisor"](https://snyk.io/advisor/golang) que mide el grado de "health" o salud de una herramienta dado su estado de mantenimiento.
-- Estándar: en caso de que el anterior factor no sea claramente decisivo, se optará por elegir la herramienta que cumple el estándar del lenguaje elegido.
+- Herramienta propia del lenguaje: en caso de que el anterior factor no sea claramente decisivo, se optará por elegir de entre las herramientas la que sea propia del lenguaje.
```
Estás cambiando algo ambiguo, "que cumple el estándar" por algo todavía más ambiguo e impreciso, "propia". Si no formulas un criterio correctamente es imposible que tomes una decisión razonable con ese criterio, porque ese criterio va a significar, como Humpty Dumpty, lo que tú quieras que signifique.
Os agradecería también que no jugárais de esta forma con las palabras para "resolver" un comentario. De veras que no es la mejor forma de emplear nuestro tiempo.

JJ commented on docs/biblioteca_aserciones.md:
```diff
@@ -19,8 +19,12 @@ Es actualizado con mucha constancia desde 2014. En el último mes se pueden obse
 ### [Paquete gotest.tools/assert](https://github.com/gotestyourself/gotest.tools)
 
-Se trata de una librería moderna de aserciones entre las herramientas de gotest.tools, pero tiene un mantenimiento un poco inestable. Prueba de ello es que solo se han realizado 23 commits este mes en un intervalo de 3 días seguidos. No existe una entrada para puntuar la herramienta en snyk Advisor, por lo que no se puede juzgar mediante este parámetro.
+Se trata de una librería moderna de aserciones entre las herramientas de gotest.tools, pero su mantenimiento es pobre. Se han realizado 23 commits este mes, lo cual es bastante inferior en comparación con el resto de herramientas. No existe una entrada para puntuar la herramienta en snyk Advisor, por lo que no se puede juzgar mediante este parámetro.
```
Sigue sin tener sentido. De hecho, se parece peligrosamente a una traducción automática de una respuesta genérica de una AI generativa. Ya he comentado en el grupo de Telegram varias veces lo que ocurre si hay sospechas en el uso de este tipo de herramientas: lo primero, que se deja de corregir. Y se advierte de que no vuelva a suceder.

JJ reviewed (commented):
Aparte de lo comentado, lee cuidadosamente los errores comunes para el resto de lo que se ha enviado.

abbonno committed:
Corrige juicio a la biblioteca de aserciones gotest.tools/assert #20

abbonno committed:
Corrige formulación criterios #20 #21

abbonno commented:
> Aparte de lo comentado, lee cuidadosamente los errores comunes para el resto de lo que se ha enviado.

De acuerdo, le daré una vuelta antes de pedir revisión.

lmchaves commented on docs/test_runner.md:
```diff
@@ -0,0 +1,34 @@
+
+## SELECCIÓN FINAL
+
+Siguiendo los criterios elegidos anteriormente, escogeremos usar el paquete de la biblioteca estándar de golang: **Go testing package**.
```
Al igaul que con la biblioteca seleccionas go testing package, pero or ejemplo Ginkgo tambien cumple con uno de los cirterios (ya que no tiiene info de Snyk Advisor)

lmchaves commented on docs/biblioteca_aserciones.md:
```diff
@@ -0,0 +1,30 @@
+
+## SELECCIÓN FINAL
+
+Siguiendo los criterios se optará por emplear el paquete de la biblioteca estándar de golang: **Go testing package**. La propia herramienta de testing Golang ofrece la posibilidad de realizar validaciones, suficientes para realizar la función de las aserciones.
```
Aqui seleccionas Go testify package, pero no usas ninguno de los criterio para validar su eleccion.

lmchaves reviewed (commented):
Buenas he realizado una serie de puntualizaciones en la seleccion de las herramientas, para que peudas superar el objetivo lo antes posible.

abbonno committed:
Añade verificación del requisito de mantenimiento para go testing #20 #21

abbonno committed:
Revert "Añade verificación del requisito de mantenimiento para go testing #20 #21"

This reverts commit af3ac9db819ae36ebf8f6ad8f7cd04c7b92d5a69.

abbonno committed:
Añade verificación del requisito de mantenimiento para go testing #20

abbonno committed:
Añade verificación del requisito de mantenimiento para go testing #21

abbonno commented:
@JJ revisión

JJ commented:
No parece que lo haya aprobado @lmchaves ¿Te parece que esperemos a su revisión y su decisión final para una segunda revisión?

lmchaves commented on docs/biblioteca_aserciones.md:
```diff
@@ -0,0 +1,30 @@
+
+## SELECCIÓN FINAL
+
+Siguiendo los criterios elegidos anteriormente se optará por emplear el paquete de la biblioteca estándar de golang: **Go testing package**. La propia herramienta de testing Golang ofrece la posibilidad de realizar validaciones, suficientes para realizar la función de las aserciones.
```
Es una simple duda, en el segundo criterio decides usar las herramientas nativas, solo si el criterio de Snyk advisor no es suficiente, y mi duda vien a porque escoges por ejemplo go testing package sobre gomega que es la herramienta que descirbes con mayor puntuación.

lmchaves reviewed (commented):
No se mucho de Go, pero basándome en lo que has escrito y los criterios que planteaste, me surge al duda sobre la selección.

abbonno committed:
Corrige snyk advisor de go testing package #20

abbonno committed:
Corrige snyk advisor de go testing package #21

lmchaves reviewed (commented):
Ahora entiendo mejor la elección pero, conforme a lo que ha dicho o JJ en telegram, al resolver los comentarios, hay que hacer los cambios en **la líena comentada**, para en este caso la elección de la herramienta quede claro

abbonno committed:
Agrega explicación snyk advisor a la selección final #20 #21

abbonno commented:
> Ahora entiendo mejor la elección pero, conforme a lo que ha dicho o JJ en telegram, al resolver los comentarios, hay que hacer los cambios en **la líena comentada**, para en este caso la elección de la herramienta quede claro

Tienes razón, lo acabo de modificar.

abbonno commented:
@JJ

JJ commented on internal/fecha_test.go:
```diff
@@ -0,0 +1,88 @@
+	}
+}
+
+func TestSumarMes(t *testing.T) {
```
Tienes 88 líneas de código en este fichero. Este test **no tiene absolutamente nada que ver con el usuario**. ¿Por qué razón ponéis para revisar código que no tiene nada que ver con la lógica principal de negocio ni la historia de usuario? ¿Qué es para vosotros un producto mínimamente viable?

JJ commented on internal/fecha.go:
```diff
@@ -0,0 +1,73 @@
+	return f
+}
+
+func (f Fecha) String() string {
```
¿Qué problema resuelve esta función?

JJ commented on internal/fecha.go:
```diff
@@ -0,0 +1,73 @@
+	return Fecha{Dia: dia, Mes: mes, Anio: anio}, nil
+}
+
+func EsBisiesto(anio int) bool {
```
Esto lo has incluido en el commit [0a5a5a0](https://github.com/abbonno/healthScheduler/commit/0a5a5a02423f350d9d87f0f68087600969c88208) que resuelve el issue #28 
1. Ese issue no plantea un problema. Hay una forma muy simple de saber si un issue plantea un problema o no. ¿Me sirve para algo o es algo que estoy poniendo porque el profe me lo ha pedido pero realmente no me sirve para nada?
2. El mensaje de commit dice [Crea clase Fecha y los test pertinentes](https://github.com/abbonno/healthScheduler/commit/0a5a5a02423f350d9d87f0f68087600969c88208). Por un lado dice una cosa obvia, que es que cre la clase Fecha, aunque no explica por qué no se reabre #9 ni por qué estamos cambiando el modelo (que se hizo en el milestone anterior) ni para el caso, por qué en el PMV que se aceptó en el objetivo 2, no se considera que es imprescindible.

JJ reviewed (changes requested):
El que todo lo que se pide en la asignatura sea útil para desarrollar y aportar valor al cliente puedes creértelo o no, eso ya es cosa de cada uno. Si no te lo crees puedes superar un objetivo, dos objetivos o tres, quizás con un poco de suerte o con algún despiste mío. No vas a llegar mucho más allá, sin embargo. Porque según vas avanzando en los objetivos, todo lo que ves son "tonterías que hay que hacer para escribir código que es lo que de verdad importa".
Y no, no importa el código. El código lo pude escribir una IA. Lo que no puede hacer una IA es plantearse problemas, sopesar diferentes opciones para resolverlos, y finalmente generare el código que las resuelva de una forma mínima y satisfactoria. Eso no lo puede hacer.
Os rogaría que tratárais de aprender lo que no es capaz de hacer el código, porque *eso* es lo que os va a dar trabajo. No crear un issue que diga "hay que crear una fecha" y decirle a ChatGPT "hazme una fecha" y generar un total de unas 150 líneas de código que tengo yo que revisar y que, cualquier cosa que haga para revisarlas, no te va a aportar ningún valor porque seguirán siendo "las tonterías que hay que hacer para escribir código"
Hay dos formas para solucionar esto. Cambias esta línea o la otra, y nos metemos en un bucle donde vamos viendo en 10 o 12 iteraciones una forma de reconducir esto por un camino en el que se pueda pasar el objetivo. Es posible que para esas fechas sea mayo de 2026, pero se llegará.
Hay otra forma. Cambiar de actitud. Tratar de entender lo que se pide y por qué. Y ver que realmente, por ese camino, es más fácil avanzar y llegar a la meta antes del 7 de febrero. Puedes hacer lo que quieras, es decisión tuya, y cualquier decisión será legítima. Cuando lo hagas, pídeme revisión de nuevo, por favor.

abbonno committed:
Aborda la fecha desde otra perspectiva #24

La modificación del modelo ha sido un error y no debería haber abordado el tema de la fecha de esa manera, no aporta valor a la lógica del milestone 1 ni forma parte del pmv

abbonno committed:
Elimina la clase Fecha

Nos deshacemos de ella porque no forma parte del pmv

abbonno committed:
Para reescribir los issues closes #22 #24 #25 #26 #27 #28

abbonno committed:
Crea función GenerarPlanAnual #30

Para obtener el plan deberemos generarlo a través de los datos que se nos ofrezca por lo que se creará la función en base a los tests siguiendo TDD

abbonno committed:
Crea TestAreaSinOcupar #31

Este test comprueba si hay áreas sin ocupar y las notifica, en lo que será un error

abbonno committed:
Crea TestAreaOcupadaDoblemente #32

Este test comprueba si hay más de una persona asignada a un área al mismo tiempo, que indica sobreocupación, y lo notifica

abbonno committed:
Modulariza comportamiento de los tests #31 #32

Aplicamos el principio DRY para evitar repetición de código creando el dato Ocupación que guarda el número de enfermeros en cada área en cada momento

abbonno committed:
Completa función GenerarPlanAnual #30

Tras completar los tests a tener en cuenta, completamos la función de creación del plan siguiendo el TDD

abbonno committed:
Crea tests para visualización turno #33

Se establecen las bases para considerar correcta la solución a aplicar para que los enfermeros comprueben sus turnos

abbonno committed:
Crea función MostrarTurnosEnfermero #33

Se crea la función que muestra al enfermero el área al que debe acudir

abbonno committed:
Soluciona problema de representación fecha #34

Al incluir una hora indeterminada usando time.Now() las igualdades entre fechas eran erróneas, se opta por crear una función que normalice las fechas dejando la hora en 0

abbonno commented:
@JJ revisión.
He modificado los issues para tratarlos desde el punto de vista de un problema a descubrir cómo resolver antes que una tarea con obligada solución. En la construcción del código he tratado de seguir la metodología de Test Driven Design para crear el código a partir de lo que se debe cumplir en los tests y no al contrario. He intentado aplicar al máximo el principio DRY. Los tests tratan de ser todo lo unitarios que pueden y, a pesar de que haya bucles anidados, estos son imprescindibles ya que tratamos de recorrer maps de tres dimensiones. Además, el código trata de hacer sola y únicamente aquello para lo que ha sido ideado.

JJ commented:
> @JJ revisión. He modificado los issues para tratarlos desde el punto de vista de un problema a descubrir cómo resolver antes que una tarea con obligada solución. En la construcción del código he tratado de seguir la metodología de Test Driven Design para crear el código a partir de lo que se debe cumplir en los tests y no al contrario. He intentado aplicar al máximo el principio DRY. Los tests tratan de ser todo lo unitarios que pueden y, a pesar de que haya bucles anidados, estos son imprescindibles ya que tratamos de recorrer maps de tres dimensiones. Además, el código trata de hacer sola y únicamente aquello para lo que ha sido ideado.

En informática siempre hay varias maneras de hacer las cosas. Si dices que algo es "imprescindible" es que simplemente no has explorado otras maneras.

JJ commented on internal/turnos-asignados.go:
```diff
@@ -15,17 +16,48 @@ const (
 	Nombre TipoTurno
 	Area   string
-	Fecha  time.Weekday
+	Fecha  time.Time
```
Todavía no has mirado por qué has hecho esto, pero es el tipo de cosas que a un revisor le hace pararse y decir qué está pasando aquí. Para empezar, es que si has cambiado el tipo, ¿no se te ha ocurrido cambiar el nombre de la variable? ¿Que un `time.Time`  se llame `Fecha` no te choca?
Pero para seguir el problema es el habitual. Es muy posible que este cambio específico haya resuelto un problema específico planteado correctamente en un issue. Si ha sido así, enhorabuena. Pero si no es así, *es incorrecto*.

JJ reviewed (commented):
De veras que lo siento, pero no voy a seguir revisando en esta vuelta después de los comentarios que te he hecho en diferentes issues. No parece que estés llevando a cabo la metodología que se enuncia en el guión.
* Estás creando issues que no representan problemas, y esto [después de lo comentado](https://github.com/abbonno/healthscheduler/pull/23#pullrequestreview-2567705523). Si todo lo que entendéis cuando hago una revisión detallada en la que invierto una buena cantidad de tiempo es "voy a hacer otra vez lo mismo para que tenga que volver a invertir la misma cantidad de tiempo", no sé si merece la pena hacerlo. Después de decirte que no había que crear issues que dijeran "hay que crear una fecha" sigue habiendo uno que dice "como crear un plan" que fue creado antes de eso, pero sobre el cual sigue habiendo código presente, código que sigue siendo inevaluable. 
* Lo que has hecho es cerrar ese issue como *completado*. No está completado, simplemente es incorrecto. Pero si es incorrecto, todo el código que se ha creado lo es, y la forma correcta de hacerlo es borrar ese código, indicando en el issue donde se hace "closes #xxx", y comenzar de nuevo, *con issues nuevos* Si no al final tenemos una serie de parches que sólo hacen todo más difícil de evaluar.
Voy a seguir corrigiendo en todo caso por si pudiera añadirte algún valor, pero si no se sigue una metodología correcta *desde el principio* es imposible que se haga así.

JJ reviewed (changes requested):
Por favor, hasta que no resuelvas *todos* los comentarios con cambios en la línea indicada, cambios que además resuelvan un problema especificado claramente en un issue, no solicites una nueva revisión.

abbonno committed:
Closes #30; closes #31; closes #32; closes #33; closes #34; closes #35 rehacer issues

abbonno committed:
Crea test para comprobar que se obtiene el empleado como solución al issue #39

Para determinar como correcta la solución al issue, un empleado debe poder comprobar el área al que asistir dados su nombre y la fecha y el turno en el que asiste

abbonno committed:
Crea función ObtenerAreaEmpleado como solución a #39, acorde a los tests

abbonno committed:
Rehace el cambio realizado a la fecha y aclara el nombre

No hay necesidad en un principio de cambiar la fecha, por lo que se devuelve a su estado original y se cambia el nombre por uno más entendible

abbonno commented:
@JJ revisión.

abbonno committed:
closes #39 elimina cambios respectivos a issue erróneo

El issue no define un problema por lo que el código creado para el mismo está mal
